### ISSUE TITLE
bug: `kflex ctx` does not work after that context is renamed

### ISSUE THREAD

**Issue by MikeSpreitzer** [kubestellar/kubestellar#2091]
### Describe the bug

I am debugging `deploy-on-k3d.md` and having trouble with the kubeflex hosting context after it is renamed. Here is a demonstration of the problem, showing that `kflex ctx` switches the current context to the original hosting context name regardless of any renaming that it has been through since then.

```shell
mspreitz@mjs13 testks % kubectl config get-contexts | fgrep \*
*         kubeflex                                                                               k3d-kubeflex                                     admin@k3d-kubeflex                                                                

mspreitz@mjs13 testks % kubectl config rename-context kubeflex kf
Context "kubeflex" renamed to "kf".

mspreitz@mjs13 testks % kubectl config current-context
kf

mspreitz@mjs13 testks % kflex ctx
✔ Checking for saved initial context...
✔ Switching to initial context...

mspreitz@mjs13 testks % kubectl config current-context
kind-kubeflex

mspreitz@mjs13 testks % kubectl config get-contexts | grep kubeflex
          kf                                                                                     k3d-kubeflex                                     admin@k3d-kubeflex                                                                
```

### Steps To Reproduce

See above.

### Expected Behavior

Not sure.

**Comment by MikeSpreitzer** [comment]
@pdettori 
@eaepstein

**Comment by pdettori** [comment]
For background, here is some info on how the initial context is maintained in the kubeconfig file: https://github.com/kubestellar/kubeflex/blob/main/docs/users.md#initial-context

Since kflex cannot know that that initial context has been renamed by another command (in  this case kubectl), I think that there are couple of possible (non exclusive) approaches around this issue:

1. Document that if the user renames the context for the KubeFlex hosting cluster, will need also to manually edit the Kubeconfig file to set the value of `kflex-initial-ctx-name` to the renamed context name.

2. Add a new subcommand to kflex to allow renaming context so that it automatically does the update of `kflex-initial-ctx-name`. 

wdyt ?

**Comment by MikeSpreitzer** [comment]
I do not like (1) because it is very tedious and very fragile. I do not like (2) because it is fragile. I think that there may be more possibilities. Could we make `kflex ctx` search the kubeconfig and find the right context by looking at attributes of all the contexts? Is there an extensions field in a kubeconfig "cluster"?

**Comment by pdettori** [comment]
It might perhaps be possible to come up with some heuristic to figure out the difference between a context for a kubeflex control plane vs. context for "real" kube clusters. But if there is more than one context for a "real" cluster (e.g. kind-kubeflex, cluster1, cluster2) I do not know how it would be possible to figure this out.  Another possibility could be to store in preferences not only the original context name but also the name of the `user` and of the `cluster` that were used in the original context and use those to identify the re-named context. Would that work ?

**Comment by ezrasilvera** [comment]
Why does KF need to know the CP(s) "contexts" ?   Isn't it the client that use KF that are using the context to access specific CP ?

**Comment by pdettori** [comment]
This is about UX with the kflex CLI. The `kflex ctx` command is used to automatically switch back to the hosting cluster context. It is also used when users does repeated commands such as `kflex create <cp-name>`, where the CLI automatically switches back to the hosting cluster context so it can create the CP object in the hosting cluster. The alternative is to have the user to manually do the extra steps (e.g. do `kflex ctx <name of hosting cluster context` before each `kflex create` or `kflex delete` command.

**Comment by ezrasilvera** [comment]
@pdettori I think this is a little  risky in any case. I always thought that solution that change the context implicitly (like KIND or KCP) is good for simple "playing" but become a potential issue on a more complex scenarios. 
I still don't understand why KF need to change context in order  to push the CP object

**Comment by pdettori** [comment]
CP objects are only available on the hosting cluster (e.g. kubeflex cluster). The kflex CLI is using the Kubeconfig and whatever context is set there to communicate with the hosting cluster.

**Comment by MikeSpreitzer** [comment]
@pdettori: remember that, as you said, the problem here is identifying the hosting cluster context; it is not identifying control plane contexts. Remember that a general user's kubeconfig has contexts for many "real" clusters (e.g., mine does), so we can not plan to identify the hosting cluster context by simply expecting that only one context is NOT for a control plane.

Yes, I just mixed up contexts and clusters, but I hope my point is clear.

But it is worth remembering that renaming a context does not rename a cluster or a user. So perhaps remembering the cluster and user names associated with the hosting context would be a possible approach.

Another idea that may help a lot is: allow `kflex init` to be told what to name the hosting cluster context, so that the user is not motivated to rename it later.

**Comment by ezrasilvera** [comment]
@pdettori the hosting cluster is where the KF controllers are running right ?  So in theory we can try to `get` the controllers on all clusters pointed by the contexts in the kubeconfig and use the one where we find the controllers.   We can even code this into a `kflex get hosting-ctx`  or something

**Comment by eaepstein** [comment]
On the bare metal machines I have been running k3s on, after cluster create the only kube config context is "default".
The initial uses of kflex are:
1. kflex init
2. kflex create its1 -t host

At this point kflex-initial-ctx-name is set to its1.
My suggestion for improvements would be:
- on kflex init, set kflex-initial-ctx-name to the name of the first cluster in kube config
- add a new command: kflex set-default-ctx  some-existing-context
- add another command to list all contexts: kflex list

**Comment by pdettori** [comment]
@eaepstein it makes sense to me, but for the first point I think it's best to set `kflex-initial-ctx-name` to the name of of the current context used to run `kflex init`. The first context in the kubeconfig could be for a different cluster not used for installing kubeflex.  I would also add a check for the existence of the initial context (after the re-naming) when switching to initial context, which would throw an error if not found.

In summary:

1. on `kflex init`, set `kflex-initial-ctx-name` to the name of the current context
2. add a new command: `kflex set-initial-ctx <some-existing-context>`
3. add another command to list all contexts: `kflex ctx list`, showing which one is the initial context and which one is the default context
4. Throw error when kflex tries to switch to an initial content that does not exist

@MikeSpreitzer @ezrasilvera - would that work ?

**Comment by eaepstein** [comment]
Completely agree with @pdettori summary above.

**Comment by pdettori** [comment]
We could have `kflex ctx-list` to avoid that trap. I thought also about `kflex ctx --list` but that does not sound right as list is a command not an option, and a list command can have also options.

**Comment by MikeSpreitzer** [comment]
The grammar problem was created by using `kflex ctx $name` to mean "switch to the context named \$name". The more usual approach (e.g., the cobra documentation says this) would be to use a command like `kflex ctx use $name` to switch contexts, leaving `kflex ctx list`, `kflex ctx create`, and so on available.

**Comment by MikeSpreitzer** [comment]
BTW, I think the critical concept here is not _initial_ cluster but rather _hosting_ cluster.

**Comment by pdettori** [comment]
@MikeSpreitzer something I did not know and found just by mistake:

```
$ kubectl ctx --help
USAGE:
  kubectl ctx                       : list the contexts
  kubectl ctx <NAME>                : switch to context <NAME>
  kubectl ctx -                     : switch to the previous context
  kubectl ctx -c, --current         : show the current context name
  kubectl ctx <NEW_NAME>=<NAME>     : rename context <NAME> to <NEW_NAME>
  kubectl ctx <NEW_NAME>=.          : rename current-context to <NEW_NAME>
  kubectl ctx -d <NAME> [<NAME...>] : delete context <NAME> ('.' for current-context)
                                  (this command won't delete the user/cluster entry
                                  that is used by the context)
  kubectl ctx -u, --unset           : unset the current context

  kubectl ctx -h,--help             : show this message
  ```
  
  so it looks like `kubectl ctx <context>` has a similar behavior to `kflex ctx <cp>`. Should we try to align to this kubextl subcommand (also for other features) rather than trying  to do `kflex ctx use <cp>` (maybe more cobra-friendly but less user friendly as users hate to type more)

**Comment by clubanderson** [comment]
Closing as likely fixed. The kflex ctx command and context handling have been improved significantly since April 2024. The current scripts use proper context management. If this issue persists with the current version, please open a new issue with reproduction steps.
